Implement join editors command

initialshl opened the pull request:
Fixes #20279

This command joins the editor group with the previous group (except for the first group which joins with the next group).

The order of editors in the joining group is preserved when joining with the target group. 

If an editor is opened in both the joining and target group, only one will be kept, in the same position as in the joining group.

![vscode-join-editors](https://cloud.githubusercontent.com/assets/19281538/23820407/ab9b9302-0652-11e7-9492-c9767ea637e4.gif)

msftclas commented:
@initialshl,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

bpasero commented on src/vs/workbench/browser/parts/editor/editor.contribution.ts:
```diff
@@ -331,6 +331,7 @@ registry.registerWorkbenchAction(new SyncActionDescriptor(CloseEditorsInGroupAct
 registry.registerWorkbenchAction(new SyncActionDescriptor(CloseOtherEditorsInGroupAction, CloseOtherEditorsInGroupAction.ID, CloseOtherEditorsInGroupAction.LABEL, { primary: null, mac: { primary: KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.KEY_T } }), 'View: Close Other Editors', category);
 registry.registerWorkbenchAction(new SyncActionDescriptor(CloseEditorsInOtherGroupsAction, CloseEditorsInOtherGroupsAction.ID, CloseEditorsInOtherGroupsAction.LABEL), 'View: Close Editors in Other Groups', category);
 registry.registerWorkbenchAction(new SyncActionDescriptor(SplitEditorAction, SplitEditorAction.ID, SplitEditorAction.LABEL, { primary: KeyMod.CtrlCmd | KeyCode.US_BACKSLASH }), 'View: Split Editor', category);
+registry.registerWorkbenchAction(new SyncActionDescriptor(JoinEditorsAction, JoinEditorsAction.ID, JoinEditorsAction.LABEL), 'View: Join Editors', category);
```
Should this maybe be more descriptive to what it actually does? `Join Editors of two Groups`?

bpasero commented on src/vs/workbench/browser/parts/editor/editorActions.ts:
```diff
@@ -106,6 +106,71 @@ export class SplitEditorAction extends Action {
+		@IWorkbenchEditorService private editorService: IWorkbenchEditorService,
+		@IEditorGroupService private editorGroupService: IEditorGroupService
+	) {
+		super(id, label, 'join-editors-action');
```
You do not need the CSS class here unless it would show up as icon, so I suggest to remove it.

bpasero commented on src/vs/workbench/browser/parts/editor/editorActions.ts:
```diff
@@ -106,6 +106,71 @@ export class SplitEditorAction extends Action {
+	}
+
+	public run(context?: IEditorContext): TPromise<any> {
+		let editorToJoin: IEditor;
```
It looks like you only need `position` and `input` from the `IEditor`. If so, I suggest to completley bypass the `editorService` API and do all of this from the `stacks model` that provides all you need (`inputs`, `groups`, a way to convert from a `group` to a `position`).

The `stacks model` is really the truth of what is going up in the 1-3 `editor groups` and you should be able to do everything with it. The more heavy `IEditor` is the actual instance of control that hosts inputs.

bpasero commented on src/vs/workbench/browser/parts/editor/editorActions.ts:
```diff
@@ -106,6 +106,71 @@ export class SplitEditorAction extends Action {
+
+		// If an editor exists in joining group and target group, only the editor in the joining group is kept.
+		if (targetPosition < toJoinPosition) {
+			toJoinEditors.forEach(e => this.editorGroupService.moveEditor(e, toJoinPosition, targetPosition, targetGroup.count));
```
This is a bit crazy to do because for each editor that gets moved it will be opened in the target group causing an editor switch. It would be much smoother if the move would happen with the editors opening `inactive` on the target group and only the last one would actually be `active`. There is an option that can be set when opening an editor called `inactive` that I think we should wire into the moveEditor call so that we can optimize for this (for the `openEditor` call [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/editorPart.ts#L866)).

initialshl commented:
@bpasero Thanks for the review!
I have refactored some parts and also used the inactive option, and added the new gif in the description of this PR.

> do we want more commands that allow to "Split Editors"?

Indeed, `Split Editors` would be a useful and convenient command, I can see myself using it a lot!

bpasero commented on src/vs/workbench/browser/parts/editor/editorPart.ts:
```diff
@@ -808,9 +808,9 @@ export class EditorPart extends Part implements IEditorPart, IEditorGroupService
-	public moveEditor(input: EditorInput, from: EditorGroup, to: EditorGroup, index?: number): void;
-	public moveEditor(input: EditorInput, from: Position, to: Position, index?: number): void;
-	public moveEditor(input: EditorInput, arg2: any, arg3: any, index?: number): void {
+	public moveEditor(input: EditorInput, from: EditorGroup, to: EditorGroup, index?: number, inactive?: boolean): void;
```
I suggest to put `index` and `inactive` into a new optional options bag (`IMoveOptions`).

bpasero commented on src/vs/workbench/services/group/common/groupService.ts:
```diff
@@ -106,9 +106,10 @@ export interface IEditorGroupService {
 	 */
-	moveEditor(input: IEditorInput, from: IEditorGroup, to: IEditorGroup, index?: number): void;
-	moveEditor(input: IEditorInput, from: Position, to: Position, index?: number): void;
+	moveEditor(input: IEditorInput, from: IEditorGroup, to: IEditorGroup, index?: number, inactive?: boolean): void;
```
`IMoveOptions`

bpasero commented on src/vs/workbench/browser/parts/editor/editorActions.ts:
```diff
@@ -106,6 +106,67 @@ export class SplitEditorAction extends Action {
+		const fromGroupTotalCount = fromGroupEditors.length;
+
+		// If an editor exists in both groups, only the editor in the joining group is kept
+		if (toPosition < fromPosition) {
```
We still need to improve this: Now we start to move editors from the group to another group starting with the first editor of that group, going to the last. This is bad when the first editor of the group is also the active editor because after each editor is moved, the group will show the next editor as active.

As such, please first move all the inactive editors of the group and only at last the active editor. This will again prevent many unwanted editor change events being fired.

initialshl commented:
@bpasero Thanks for the feedback. I overlooked the `inactive` option that can be used to control the switching of the active editor. It appears that `preserveFocus` has to be set too, so that the `toGroup` does not get activated when moving inactive editors.

And regarding the possibility of adding a new command `SplitEditors`, I would be happy to contribute another PR for that if it is good 👍

bpasero commented:
@initialshl yes, something like "View: Split Editors into Two Groups" 👍 

This PR looks nice now, thanks. I will merge it in tomorrow.

kounelios13 commented:
is there any key combination available?

bpasero commented:
@kounelios13 just assign it via keybindings settings:

![image](https://user-images.githubusercontent.com/900690/35193829-e04d2f94-fea8-11e7-9073-2e33e5cccdf1.png)

kounelios13 commented:
Hello.I've noticed something.If I have more than 2 editors open when I execute  this command it won't join all of them.Is there any way to achieve this behavior?

bpasero commented:
Works for me, even with 3 editor groups open.

kounelios13 commented:
@bpasero If I have 3 editors open it will merge two of them and then it leaves me with 2 editors open

bpasero commented:
@kounelios13 what would you expect to happen?

kounelios13 commented:
@bpasero I would expect that if I have let's say 5 editors open after executing the join command it would merge them all into 1.

Isn't that the purpose of this command or have I misunderstood it?

bpasero commented:
@kounelios13 as the name of that action implies, it will only merge the editors of 2 groups, not all groups.

kounelios13 commented:
Sorry my mistake.
